ktor-server/ktor-server-plugins/ktor-server-status-pages/jvmAndNix/src/io/ktor/server/plugins/StatusPages.kt

```diff
@@ -7,136 +7,95 @@ package io.ktor.server.plugins
+     */
+    public fun status(
+        vararg status: HttpStatusCode,
+        handler: suspend (ApplicationCall, HttpStatusCode) -> Unit
+    ) {
+        status.forEach {
+            statuses[it] = handler
         }
     }
+}
 
-    private suspend fun interceptCall(context: PipelineContext<Unit, ApplicationCall>) {
-        try {
-            coroutineScope {
-                context.proceed()
-            }
-        } catch (exception: Throwable) {
-            val handler = findHandlerByValue(exception)
-            if (handler != null && context.call.response.status() == null) {
-                context.handler(exception)
-                finishIfResponseSent(context)
-            } else {
-                throw exception
-            }
-        }
-    }
+@SharedImmutable
+public val StatusPages: ApplicationPlugin<Application, StatusPagesConfig, PluginInstance> = createApplicationPlugin(
+    "StatusPages",
+    { StatusPagesConfig() }
+) {
+    val statusPageMarker = AttributeKey<Unit>("StatusPagesTriggered")
 
-    private fun findHandlerByValue(cause: Throwable): HandlerFunction? {
+    val exceptions = HashMap(pluginConfig.exceptions)
+    val statuses = HashMap(pluginConfig.statuses)
+
+    fun findHandlerByValue(cause: Throwable): HandlerFunction? {
         val key = exceptions.keys.find { cause.instanceOf(it) } ?: return null
         return exceptions[key]
     }
 
-    /**
-     * Plugin installation object
-     */
-    public companion object Plugin : ApplicationPlugin<ApplicationCallPipeline, Configuration, StatusPages> {
-        override val key: AttributeKey<StatusPages> = AttributeKey("Status Pages")
-
-        override fun install(pipeline: ApplicationCallPipeline, configure: Configuration.() -> Unit): StatusPages {
-            val configuration = Configuration().apply(configure)
-            val plugin = StatusPages(configuration)
-            if (plugin.statuses.isNotEmpty()) {
-                pipeline.sendPipeline.intercept(ApplicationSendPipeline.After) { message ->
-                    plugin.interceptResponse(this, message)
-                }
-            }
-            if (plugin.exceptions.isNotEmpty()) {
-                pipeline.intercept(ApplicationCallPipeline.Monitoring) {
-                    plugin.interceptCall(this)
-                }
-            }
-            return plugin
-        }
+    onCallRespond { call: ApplicationCall, body: Any ->
+        if (call.attributes.contains(statusPageMarker)) return@onCallRespond
+
+        val status = when (body) {
+            is OutgoingContent -> body.status
+            is HttpStatusCode -> body
+            else -> null
+        } ?: return@onCallRespond
+
+        val handler = statuses[status] ?: return@onCallRespond
+        call.attributes.put(statusPageMarker, Unit)
+        handler(call, status)
     }
-}
 
-private typealias HandlerFunction = suspend PipelineContext<Unit, ApplicationCall>.(Throwable) -> Unit
+    on(CallFailed) { call, cause ->
```

Comment:
Even though I really like the simplicity of `on(BlahBlah)` maybe `onEvent(BlahBlah)` would be more explanative?
Not entirely sure though, I'm tempted to use the shorter option anyway. But let's also discuss later with the team

Reply:
I like the parity with other handlers(like `onCall` or `onCallReceive`), will add to the discussions list